torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,432 @@
+          batch_data.batch_data.emplace_back(std::move(data[i]));
+        }
+        remaining_size -= example_count;
+      }
+    }
+
+    // If we still have data remaining after filling the last pushed batch, add
+    // them to the queue too.
+    while (remaining_size > 0) {
+      UnwrappedBatchType current_batch;
+
+      // Allocate the batch memory ahead of time.
+      current_batch.reserve(batch_size_);
+
+      auto example_count = std::min(remaining_size, batch_size_);
+      auto batch_example_indices = example_sampler_.next(example_count);
+      AT_ASSERT(
+          batch_example_indices &&
+          batch_example_indices.value().size() == example_count)
+      BatchRequestType indices = batch_example_indices.value();
+      for (size_t i : indices) {
+        current_batch.emplace_back(std::move(data[i]));
+      }
+      remaining_size -= example_count;
+      UnwrappedBatchData<UnwrappedBatchType> batch_data(
+          std::move(current_batch));
+      batch_queue_.push(std::move(batch_data));
+    }
+    this->total_example_count_in_queue_ += data_size;
+
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    this->remaining_chunk_count_--;
+
+    lock.unlock();
+    cvr_.notify_all();
+  }
+
+  /// Push exceptions throwed during preloading into batch queue. Called from
+  /// the ChunkDataSet worker threads.
+  void add_chunk_data(size_t index, std::exception_ptr e_ptr) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cvw_.wait(lock, [this] {
+      // stop loading if we have preloaded enough data.
+      return this->total_example_count_in_queue_ < queue_depth_;
+    });
+    UnwrappedBatchData<UnwrappedBatchType> batch_data(e_ptr);
+    batch_queue_.push(std::move(batch_data));
+
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cvr_.notify_all(); // notify all readers.
+  }
+
+  /// count of remaining chunk to be loaded. It is initialized with the total
+  /// chunk count and it decreases when a chunk data is retrieved. When this reaches
+  /// to 0, no more chunk needs to be loaded.
+  size_t remaining_chunk_count_ = 0;
+
+  /// The batch size is needed to create batches from the chunk data. Similar to
+  /// regular dataloader where the batches are created with prefetches,
+  /// BatchDataBuffer perform the batch creation using the provided batch size.
+  size_t batch_size_ = 0;
+
+  /// count of total example stored in the queue
+  size_t total_example_count_in_queue_ = 0;
+
+  /// local cache to store example batches from loaded chunk
+  std::queue<UnwrappedBatchData<UnwrappedBatchType>> batch_queue_;
+
+  // sync batch_queue_ update.
+  std::mutex queue_mutex_;
+
+  std::condition_variable cvr_;
+  std::condition_variable cvw_;
+
+  ExampleSampler example_sampler_;
+
+  // indicator for whether an empty chunk should be ignored. When it is true, an
+  // example will throw, otherwise, this empty chunk is skipped.
+  bool ignore_empty_chunk_ = false;
```

Comment:
This is added as a precaution check. When the chunk is empty, it could be by design, but also, it could be many other reasons like the file is corrupted. The user knows better to decide whether it is harmless. This gives them the option if they want to investigate it.

Reply:
Honestly I don't get why do we even have this flag. If you want to treat empty chunks as errors, simply make sure that your `ChunkReader` never returns one, and throws instead. It only makes the implementation more complicated, with no added expressive power.

Reply:
> Empty chunks are real.

No discussion here

> We cannot expect all the readers to catch them either.

Why? Readers are implemented by people, and they can decide which behavior do they prefer. They have to be aware that the code *they wrote* can produce certain behaviors, and it's up to them to guard against them or not.

Reply:
Of course, hope we don't invent coding robots :). The problem is, someone finding an empty chunk due to data some data anomaly should not stop a training that ran for few hours. I really don't think if there is a big difference in code where we put this logic. It is a safe guard in the in our code.

Reply:
> The problem is, someone finding an empty chunk due to data some data anomaly should not stop a training that ran for few hours.

Exactly! Which is why this dataset should be written with `ignore_empty_chunk_` hardcoded to `true`, and if someone _really_ wants to treat them as errors, they are welcome to do so in their reader.

Really, there's a lot of value in providing the simplest interface possible that gets the job done. This flag is simply unnecessary, and only clutters the API.

Reply:
Ok, let's remove this and at least log that we are ignoring an empty chunk.